src/main/scala/chiseltest/backends/VPIInterface.scala

```diff
@@ -0,0 +1,427 @@
+    extends SimulatorInterface
+    with LazyLogging {
+  def cycleCount: BigInt = _currentClock
+
+  def poke(signal: String, value: BigInt): Unit = {
+    if (inputsNameToChunkSizeMap contains signal) {
+      _pokeMap(signal) = value
+      isStale = true
+    } else {
+      val id = _signalMap.getOrElseUpdate(signal, getId(signal))
+      if (id >= 0) {
+        poke(id, _chunks.getOrElseUpdate(signal, getChunk(id)), value)
+        isStale = true
+      } else {
+        logger.info(s"Can't find $signal in the emulator...")
+      }
+    }
+  }
+
+  def peek(signal: String): Option[BigInt] = {
+    if (isStale) update()
+    if (outputsNameToChunkSizeMap.contains(signal)) {
+      _peekMap.get(signal)
+    } else if (inputsNameToChunkSizeMap.contains(signal)) {
+      // Added this in case peek is called on input before poke. Did not seem to be a problem in testers
+      if (!_pokeMap.contains(signal)) {
+        _pokeMap(signal) = BigInt(0)
+        isStale = true
+        update()
+      }
+      _pokeMap.get(signal)
+    } else {
+      val id = _signalMap.getOrElseUpdate(signal, getId(signal))
+      if (id >= 0) {
+        Some(peek(id, _chunks.getOrElse(signal, getChunk(id))))
+      } else {
+        logger.info(s"Can't find $signal in the emulator...")
+        None
+      }
+    }
+  }
+
+  def step(n: Int): Unit = {
+    update()
+    (0 until n).foreach(_ => takeStep())
+  }
+
+  def finish(): Unit = {
+    mwhile(!sendCmd(SIM_CMD.FIN)) {}
+    logger.info("Exit Code: %d".format(Await.result(exitValue, Duration.Inf)))
+    dumpLogs()
+    inChannel.close()
+    outChannel.close()
+    cmdChannel.close()
+  }
+
+  /* VPI interface implementations. */
+  val (inputsNameToChunkSizeMap, outputsNameToChunkSizeMap) = {
+    val (inputs, outputs) = topPorts.partition(_.direction == firrtl.ir.Input)
+
+    /** genChunk only will gen vaild Chunk now. */
+    def genChunk(args: Port): Option[(String, Int)] = {
+      args match {
+        case Port(_, name, _, GroundType(width)) if width.asInstanceOf[IntWidth].width > 0 =>
+          //TODO: This hack needs to be replaced by something rational
+          val n = if (name.endsWith("reset")) {
+            "reset"
+          } else if (name.endsWith("clock")) {
+            "clock"
+          } else {
+            s"$topName.$name"
+          }
+          Some(n -> ((width.asInstanceOf[IntWidth].width.toInt - 1) / 64 + 1))
+        case _ =>
+          None
+      }
+    }
+
+    (ListMap(inputs.flatMap(genChunk): _*), ListMap(outputs.flatMap(genChunk): _*))
+  }
+
+  private object SIM_CMD extends Enumeration {
+    val RESET, STEP, UPDATE, POKE, PEEK, FORCE, GETID, GETCHK, FIN = Value
+  }
+
+  implicit def cmdToId(cmd: SIM_CMD.Value): Int = cmd.id
+
+  implicit def int(x: Int): BigInt = (BigInt(x >>> 1) << 1) | BigInt(x & 1)
+
+  implicit def int(x: Long): BigInt = (BigInt(x >>> 1) << 1) | BigInt(x & 1)
+
+  private var isStale = false
```

Comment:
Why do some variables have underscores and some don't?

Reply:
It's was legacy code.
I'll make it legacy and refactor in the future PR.